### PR TITLE
ConfigMap changes on update cause startup failures #1150

### PR DESCRIPTION
**Pull request by ryanemerson**

### CODE REVIEW

File: pkg/controller/infinispan/resources/config/config_controller.go [hunk 1]
```diff
@@ -74,6 +75,16 @@ func Add(mgr manager.Manager) error {
 }
 
 func (c *configResource) Process() (reconcile.Result, error) {
+	// Don't update the ConfigMap if an update will be required
+	if req, err := ispnctrl.IsUpgradeRequired(c.infinispan); req || err != nil {
+		return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCreateResource}, nil
+	}
+
+	// If an upgrade is in progress wait for it to complete before updating the ConfigMap
+	if c.infinispan.IsAnyCondititionTrue(ispnv1.ConditionUpgrade, ispnv1.ConditionStopping) {
```
Comment (dmvolod) [inline pkg/controller/infinispan/resources/config/config_controller.go:84]: If #763 wouldn't be fixed, this code cause continuous loop

Comment (ryanemerson) [reply pkg/controller/infinispan/resources/config/config_controller.go:84]: True. I wasn't aware of that issue. That issue needs to be fixed eventually, but I don't think it should stop this fix. Worst case scenario the ConfigMap retains the old data whilst `ConditionStopping` and the config-controller will loop until this is resolved.

Comment (dmvolod) [reply pkg/controller/infinispan/resources/config/config_controller.go:84]: Probably, we could validate upgrade condition again here, i.e. no Pods or Pods with DEFAULT_IMAGE different from operator configuration

File: pkg/controller/infinispan/infinispan_controller.go [hunk 2]
```diff
@@ -706,16 +719,23 @@ func (r *ReconcileInfinispan) scheduleUpgradeIfNeeded(infinispan *infinispanv1.I
...
+	if err != nil {
+		return false, err
 	}
-	return nil, nil
+	return upgradeRequired(infinispan, podList)
+}
+
+func PodList(infinispan *infinispanv1.Infinispan) (*corev1.PodList, error) {
+	podList := &corev1.PodList{}
+	if err := kubernetes.ResourcesList(infinispan.Namespace, PodLabels(infinispan.Name), podList); err != nil {
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:735]: can we just return `podList, err` without validation?

File: pkg/controller/infinispan/infinispan_controller.go [hunk 3]
```diff
@@ -683,19 +683,32 @@ func (r *ReconcileInfinispan) upgradeInfinispan(infinispan *infinispanv1.Infinis
 }
 
 func (r *ReconcileInfinispan) scheduleUpgradeIfNeeded(infinispan *infinispanv1.Infinispan, podList *corev1.PodList, logger logr.Logger) (*reconcile.Result, error) {
+	if upgrade, err := upgradeRequired(infinispan, podList); upgrade || err != nil {
+		if err := r.update(infinispan, func() {
+			logger.Info("schedule an Infinispan cluster upgrade", "pod default image", podList.Items[0].Spec.Containers[0], "desired image", consts.DefaultImageName)
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:688]: Probably image is missing, whole container will be printing

File: pkg/controller/infinispan/resources/config/config_controller.go (hunk 1 above)
Comment (ryanemerson) [reply pkg/controller/infinispan/resources/config/config_controller.go:84]: That's handled in the call to `if req, err := ispnctrl.IsUpgradeRequired(c.infinispan); req || err != nil {`

File: pkg/controller/infinispan/infinispan_controller.go (hunk 3 above)
Comment (ryanemerson) [reply pkg/controller/infinispan/infinispan_controller.go:688]: I don't understand what you mean? In order for `upgradeRequired` to return true, it's necessary for `len(podList) > 0`.

File: pkg/controller/infinispan/resources/config/config_controller.go (hunk 1 above)
Comment (dmvolod) [reply pkg/controller/infinispan/resources/config/config_controller.go:84]: Why we need additional validation ` c.infinispan.IsAnyCondititionTrue` in this case?

File: pkg/controller/infinispan/infinispan_controller.go (hunk 3 above)
Comment (dmvolod) [reply pkg/controller/infinispan/infinispan_controller.go:688]: `podList.Items[0].Spec.Containers[0]` returns whole container object, not an image

File: pkg/controller/infinispan/resources/config/config_controller.go (hunk 1 above)
Comment (ryanemerson) [reply pkg/controller/infinispan/resources/config/config_controller.go:84]: They handle two different scenarios. The first if statement is to guard against the config-controller processing Infinispan before infinispan_controller can initiate the upgrade and sets the appropriate conditions. The second if statement ensures that if a upgrade has been initiated, we don't update the Config until after the original cluster pods have been shutdown.

Comment (dmvolod) [reply pkg/controller/infinispan/resources/config/config_controller.go:84]: I'm not sure, but seems to new Pods never will have new DEFAULT_IMAGE value till upgrade has finished. Seems to second validation is bit redundant, but maybe I'm wrong

File: pkg/controller/infinispan/infinispan_controller.go (hunk 3 above)
Comment (ryanemerson) [reply pkg/controller/infinispan/infinispan_controller.go]: Gotcha! My copy and paste skills failed me there. Updated.

File: pkg/controller/infinispan/resources/config/config_controller.go (hunk 1 above)
Comment (ryanemerson) [reply pkg/controller/infinispan/resources/config/config_controller.go:84]: > Seems to second validation is bit redundant, but maybe I'm wrong

Hmm actually I think you're right, as `IsUpgradeRequired` already checks if `ConditionUpgrade` is true. I'll try removing and verify locally.

Comment (dmvolod) [reply pkg/controller/infinispan/resources/config/config_controller.go:84]: Yes, `ispnv1.ConditionStopping` a bit danger validation as it's related to graceful shutdown also. But in this case we can free to change any config, as it's not upgrade

File: pkg/controller/infinispan/infinispan_controller.go [hunk 4]
```diff
@@ -706,16 +720,20 @@ func (r *ReconcileInfinispan) scheduleUpgradeIfNeeded(infinispan *infinispanv1.I
...
+	if err != nil {
+		return false, err
 	}
-	return nil, nil
+	return upgradeRequired(infinispan, podList)
+}
+
+func PodList(infinispan *infinispanv1.Infinispan) (*corev1.PodList, error) {
+	podList := &corev1.PodList{}
+	return podList, kubernetes.ResourcesList(infinispan.Namespace, PodLabels(infinispan.Name), podList)
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:736]: seems to we need execute `err := kubernetes.ResourcesList(infinispan.Namespace, PodLabels(infinispan.Name), podList)` and then `return podList, err`
May be I'm wrong

File: pkg/controller/infinispan/resources/config/config_controller.go (hunk 1 above)
Comment (ryanemerson) [reply pkg/controller/infinispan/resources/config/config_controller.go:84]: Do we use graceful shutdown for anything other than upgrades?

Comment (dmvolod) [reply pkg/controller/infinispan/resources/config/config_controller.go:84]: Yes, sure. If anyone defines replicas=0 operator will execute graceful shutdown

File: pkg/controller/infinispan/infinispan_controller.go (hunk 4 above)
Comment (ryanemerson) [reply pkg/controller/infinispan/infinispan_controller.go:736]: I don't think it matters, as `podList` is just a pointer. The call to `ResourcesList` updates the value associated with the pointer.

File: pkg/controller/infinispan/resources/config/config_controller.go (hunk 1 above)
Comment (ryanemerson) [reply pkg/controller/infinispan/resources/config/config_controller.go:84]: Ok, in that case it should be fine to remove the `ConditionUpgrade` check as scaling the cluster back up will contain the same Infinispan version and so we can't run into this issue.

**Review by dmvolod — APPROVED** [review]

**Pull request merged**